crates/devolutions-pedm/src/db/mod.rs

```diff
@@ -14,22 +19,172 @@ mod pg;
+
+                info!(
+                    "Connecting to postgres://{user}@{host}:{port}/{dbname}",
+                    user = c.user,
+                    host = c.host,
+                    port = c.port.unwrap_or(5432),
+                    dbname = c.dbname
+                );
+                // Check if the connection works.
+                let conn = pool.get().await?;
+                conn.query_one("SELECT 1", &[]).await?;
+                drop(conn);
+                Arc::new(PgPool::new(pool))
+            }
+        };
+        info!("Successfully connected to the database");
+        Ok(Self(db))
+    }
+
+    /// Initializes the database schema if needed.
+    pub(crate) async fn init_schema(&self) -> Result<(), InitSchemaError> {
+        match self.0.get_schema_version().await {
+            Ok(version) => {
+                info!("Schema version: {version}");
+                if version == 1 {
+                    Ok(())
+                } else {
+                    Err(InitSchemaError::VersionMismatch {
+                        expected: 1,
+                        actual: version,
+                    })
+                }
+            }
+            Err(error) => {
+                if error.is_table_does_not_exist() {
+                    info!("Initializing schema");
+                    self.0.init_schema().await?;
+                    Ok(())
+                } else {
+                    Err(error.into())
+                }
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum InitSchemaError {
+    VersionMismatch { expected: i16, actual: i16 },
+    Db(DbError),
+}
+
+impl core::error::Error for InitSchemaError {
+    fn source(&self) -> Option<&(dyn core::error::Error + 'static)> {
+        match self {
+            Self::VersionMismatch { .. } => None,
+            Self::Db(e) => Some(e),
+        }
+    }
+}
+
+impl fmt::Display for InitSchemaError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::VersionMismatch { expected, actual } => {
+                write!(
+                    f,
+                    "schema version mismatch: expected version {expected}, got version {actual}"
+                )
+            }
+            Self::Db(e) => e.fmt(f),
+        }
+    }
+}
+
+impl From<DbError> for InitSchemaError {
+    fn from(e: DbError) -> Self {
+        Self::Db(e)
+    }
+}
+
 /// Abstracts database operations for backends such as Postgres or libSQL.
 ///
 /// All queries required by the application are defined here. They must be implemented by each backend.
 #[async_trait]
 pub(crate) trait Database: Send + Sync {
+    /// Returns the schema version from the `version` table.
+    async fn get_schema_version(&self) -> Result<i16, DbError>;
+
+    /// Initializes the database schema.
+    ///
+    /// This creates tables.
+    async fn init_schema(&self) -> Result<(), DbError>;
```

Comment:
question: Should we really expose the schema version to the user of the `Database` trait?

I think it’s more appropriate to define a single method called `setup` that is performing any backend-dependent setup sequence, including initializing the schema and/or performing migrations if necessary.

See how it’s done in `job-queue`/`job-queue-libsql` crates: https://github.com/Devolutions/devolutions-gateway/blob/cdbe0b7a00e0ced13c90fa4198eabe173b272140/crates/job-queue-libsql/src/lib.rs#L174-L178

Reply:
Fair enough. My thinking was to keep the exact setup logic as an implementation detail to leave more leeway, but at this point it does not really matter so we may as well avoid the logic duplication as you suggest. Certainly a better abstraction will emerge when it becomes necessary.